### PR TITLE
Backport #3661: Add twirl sources to eclipse classpath.

### PR DESCRIPTION
**Pull request by oleastre**
As playframework 2.3.x already uses Twirl, the generated source code should be added to eclipse source classpath. 
The fix is already merge in master as #3661 and this is just a backport for the current 2.3.x branch.

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @oleastre,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

**Comment by jroper** [comment]
Doesn't sbt eclipse already automatically add all the source managed directories?

**Comment by oleastre** [comment]
I just tested with the following:
```bash
$ ./activator new test play-scala
$ echo -e '\naddSbtPlugin("com.typesafe.sbteclipse" % "sbteclipse-plugin" % "3.0.0")' >> test/project/plugins.sbt
$ cd test
$ sbt eclipse
```
If I take a look at the generated .classpath file, there is no entry for the twirl generated sources.
If I import the generated project in eclipse, I get the following error:
```object index is not a member of package views.html```

Does not seem to work out of the box in eclipse.
That said, this PR is just a back port of some code that's already in play master, so I suppose the choice between patching sbt-eclipse or play has already been discussed.

**Comment by oleastre** [comment]
After further exploration, it seems that sbteclipse does not includes, by default, managed source in it's classpath. See:
https://github.com/typesafehub/sbteclipse/issues/217

**Comment by oleastre** [comment]
I did a PR in sbteclipse to add automatically source managed directories: https://github.com/typesafehub/sbteclipse/pull/242

If this one is merged, the code that adds `src_managed` in the classpath should be removed.

**Comment by benmccann** [comment]
Is this PR equivalent to the sbteclipse PR? If so, why not just change the EclipseKeys.createSrc setting here and get rid of all the addSourceDirectory craziness?

**Comment by oleastre** [comment]
The sbteclipse PR is another way to solve the same problem, and is an answer to @jroper comment.

**Comment by benmccann** [comment]
Yeah, I guess the thing I don't understand is why the code for the PR is so different than the code for the sbteclipse PR. It seems like that approach would work here just as well and might be a lot cleaner.

**Comment by jroper** [comment]
@play-pull-request-validator

**Comment by oleastre** [comment]
I also think that adding all managed sources folders in the generated classpath is a better default.

To solve this issue, playframework can probably switch to setting EclipseKeys.createSrc to All. If the team is ok with that solution, I will be happy to submit a new PR for both master and 2.3.x branch.

If sbteclipse accepts my other PR, play will then just have to remove the EclipseKeys.createSrc config.

If the play team would be more conservative, and not change the build chain in a stable release, then the current PR could be applied to the 2.3.x branch and the EclipseKeys.createSrc be changed in master.

**Comment by benmccann** [comment]
I'd probably prefer to wait for yous sbteclipse change to be merged before making any changes here. It seems like a good idea to me, but I'm curious if there's any reason they didn't do it from the beginning or if there are any problems we're not aware of. Let's just keep pinging that pull request every so often to remind them to take a look.

**Comment by oleastre** [comment]
In the mean time play 2.3.8 as been released without any fix for this issue. 
Shouldn't this backport PR just be merged and a new PR created to discuss the future solution (be it a change to EclipseKeys.createSrc default value on the sbt-eclipse side or on the playframework side) ?
Please let's have the next play release work out of the box with eclipse...

**Comment by benmccann** [comment]
I think we're pretty close on your PR at sbteclipse, so we should be able to get that done soon. I just want to figure out the bug with the duplicate classpath entries that you mentioned. Can you share how to reproduce that issue over on that thread?

**Comment by benmccann** [comment]
Alternative PR here: https://github.com/playframework/playframework/pull/4074

**Comment by benmccann** [comment]
@oleastre now that https://github.com/playframework/playframework/pull/4074 has been merged, we might be able to backport that one. I think that'd be a better way of solving this, so I'm going to close this. Thanks for bringing the issue to our attention

**Comment by benmccann** [comment]
Reopening. I think this is a better way to fix on the 2.3.x branch because it won't create a bunch of empty directories in Eclipse. On 2.4.x we'll upgrade to use a newer version of sbteclipse that doesn't have that problem after sbteclipse 247 is merged

File: framework/src/sbt-plugin/src/main/scala/PlayEclipse.scala [hunk 1]
```diff
@@ -70,14 +71,19 @@ trait PlayEclipse {
       }
     }
```
Comment (benmccann) [inline framework/src/sbt-plugin/src/main/scala/PlayEclipse.scala]: @oleastre one last comment before i merge this. could you also put `addRoutesSources` here as we had on master?

    lazy val addRoutesSources = addSourceDirectory(target in (Compile, RoutesKeys.routes))

**Comment by benmccann** [comment]
@oleastre thanks for the patience and the help on this. i know it's taken awhile, but i'm glad we're getting it right

File: framework/src/sbt-plugin/src/main/scala/PlayEclipse.scala (hunk 1 above)
Comment (oleastre) [inline framework/src/sbt-plugin/src/main/scala/PlayEclipse.scala]: I had removed addRoutesSources from the patch, as RoutesKeys.routes has been added in the 2.4 release and does not exists in 2.3.
The generated routes files are not handled by a separate sbt task but directly added in the sourceGenerator with (sourceManaged in Compile) as target path.
For the appending to sourceGenerator, see:
https://github.com/playframework/playframework/blob/2.3.x/framework/src/sbt-plugin/src/main/scala/PlaySettings.scala#L118
And the RouteFiles definition at:
https://github.com/playframework/playframework/blob/2.3.x/framework/src/sbt-plugin/src/main/scala/PlaySourceGenerators.scala#L12

As the sourceManaged folder is already added by the previous line (lazy val addSourcesManaged), there is no need to add something different for the generated route files.

**Comment by oleastre** [comment]
I also prefer to get things done in the right way, and this can take time, and I'm ready to help as much as I can.
But sometimes, I just need tools that works; that's why I'm a little bit pressing to get this done :)
Thanks for your support.

File: framework/src/sbt-plugin/src/main/scala/PlayEclipse.scala (hunk 1 above)
Comment (benmccann) [inline framework/src/sbt-plugin/src/main/scala/PlayEclipse.scala]: Ok, thanks for the clarification. That sounds fine then

**Pull request merged**